api/lql.go

```diff
@@ -18,24 +18,177 @@
 
 package api
 
+import (
+	"encoding/json"
+	"net/url"
+	"regexp"
+	"strconv"
+	"time"
+
+	"github.com/pkg/errors"
+)
+
+const (
+	reLQL                  string = `(?ms)^(\w+)\([^)]+\)\s*{`
+	LQLQueryTranslateError string = "unable to translate query blob"
+)
+
+type LQLQuery struct {
+	ID             string `json:"LQL_ID,omitempty"`
+	StartTimeRange string `json:"START_TIME_RANGE,omitempty"`
+	EndTimeRange   string `json:"END_TIME_RANGE,omitempty"`
+	QueryText      string `json:"QUERY_TEXT"`
+	// QueryBlob is a special string that supports type conversion
+	// back and forth from LQL to JSON
+	QueryBlob string `json:"-"`
+}
+
+func (q *LQLQuery) Translate() (returnError error) {
+	// if QueryText is populated; return
+	if q.QueryText != "" {
+		return
+	}
+
+	// if QueryBlob is JSON
+	var t LQLQuery
+
+	if err := json.Unmarshal([]byte(q.QueryBlob), &t); err == nil {
```

Comment:
No.  From an end user standpoint the fact that we tried to unmarshal something that wasn't JSON is irrelevant here (since we're dealing with a Blob). We would want a consistent err msg experience from this fun.

Reply:
I disagree. If we are trying to give a really amazing experience creating and running queries from the CLI, we must provide the reason why we couldn't translate a query. This is a major problem we have had where we assume that by saying `"unable to translate query blob"` the user will know what to do next. They won't. They will be confused. If we can tell them why we couldn't translate the query, we have to tell them, not hide it, this will allow them to troubleshoot the problem on their own.

I just ran a quick exercise where I am trying to run a simple query:
<img width="1420" alt="Screen Shot 2021-04-02 at 6 02 27 PM" src="https://user-images.githubusercontent.com/5712253/113462577-d3c72a80-93de-11eb-809b-3c3732f0d917.png">

I type my query:
<img width="1420" alt="Screen Shot 2021-04-02 at 6 00 40 PM" src="https://user-images.githubusercontent.com/5712253/113462607-ec374500-93de-11eb-8de4-b479eb2eccc1.png">

And, run it!
<img width="1420" alt="Screen Shot 2021-04-02 at 6 13 55 PM" src="https://user-images.githubusercontent.com/5712253/113462662-36b8c180-93df-11eb-9d49-4d7c45eda260.png">

What did I do wrong? How can I troubleshoot this problem? There is nothing....

Instead, if we bubble-up the error message to the end-user, we will get messages like:
```
ERROR unable to run LQL query: unable to translate query blob: invalid character '"' after object key:value pair
```

```
ERROR unable to run LQL query: unable to translate query blob: unexpected end of JSON input
```

```
ERROR unable to run LQL query: unable to translate query blob: invalid character ']' looking for beginning of value
```

This 👆🏽 in my opinion is more helpful. I am happy to jump into a zoom with you to talk about it further. 😈

Reply:
I couldn't agree more that we can provide better messaging.  However we must keep in mind that there are multiple query formats in play.  So if you mess up a text-based query, it doesn't matter that this doesn't parse in JSON.  If you mess up a JSON-based query it doesn't matter that it doesn't match the text-based regex.  So when there's completely arbitrary text I'm not certain how specific we can be about what you specifically messed up...

Reply:
@afiune , I think we are in agreement here that we want to improve this.  I went ahead and filed https://lacework.atlassian.net/browse/ALLY-432 to make that happen.  Let me know if you feel that this is blocking for the PR.